### PR TITLE
Add initial Sentiment Analysis tutorial to Natural Language samples

### PR DESCRIPTION
**Pull request by manshreck**
This sample will be used within the Sentiment Analysis Tutorial at:

https://cloud.google.com/natural-language/docs/sentiment-tutorial

This pull does not yet include [SECTION] tags, which I will add subsequently.

### CODE REVIEW

File: language/sentiment/README.txt [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+within the Natural Language API Documentation. A full walkthrough of this sample
+is located within the documentation.
+
+This sample is a simple illustration of how to construct a sentiment analysis
+request and process a response using the API.
+
+## Download the Code
+
+```
+git clone sso://python-dev-samples/language/sentiment/
```
Comment (jerjou) [inline language/sentiment/README.txt]: `git clone https://github.com/GoogleCloudPlatform/python-docs-samples.git`

File: language/sentiment/README.txt [hunk 2]
```diff
@@ -0,0 +1,27 @@
+# Introduction
```
Comment (jerjou) [inline language/sentiment/README.txt]: The file should have an `.md` extension

File: language/sentiment/README.txt [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+
+This sample contains the code referenced in the 
+[Sentiment Analysis Tutorial](http://cloud.google.com/natural-language/docs/sentiment-tutorial)
+within the Natural Language API Documentation. A full walkthrough of this sample
```
Comment (jerjou) [inline language/sentiment/README.txt]: First mention of the product should use the entire (verbose) name

File: language/sentiment/README.txt [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+This sample is a simple illustration of how to construct a sentiment analysis
+request and process a response using the API.
+
+## Download the Code
+
+```
+git clone sso://python-dev-samples/language/sentiment/
+```
+
+### Usage
```
Comment (jerjou) [inline language/sentiment/README.txt]: Should probably add a section for creating a virtualenv and installing the requirements. [example](https://github.com/GoogleCloudPlatform/python-docs-samples/tree/master/speech/api-client#setup)

File: language/sentiment/mixed.txt [hunk 5]
```diff
@@ -0,0 +1,20 @@
+I really wanted to love 'Bladerunner' but ultimately I couldn't get
```
Comment (jerjou) [inline language/sentiment/mixed.txt]: Can you rename the `sentiment` dir to `resources`, for consistency with the other samples

File: language/sentiment/out [hunk 6]
```diff
@@ -0,0 +1 @@
...bG9zb3BoaWNhbApleHBsb3JhdGlvbiBvZiB3aGF0IGl0IG1lYW5zIHRvIGJlIGh1bWFuIG9yIG1hY2hpbmUuIFNvbWUgb2YgdGhlIGdpem1vcwpsaWtlIHRoZSBmbHlpbmcgY2FycyBhbmQgdGhlIFZvdWlnaHQtS2FtcGZmIG1hY2hpbmUgKHdoaWNoIHNlZW1lZCB2ZXJ5CnN0ZWFtcHVuayksIHdlcmUgcXVpdGUgY29vbC4KCkkgZGlkIGZpbmQgdGhlIHBsb3QgcHJldHR5IHNsb3cgYW5kIGJ1dCB0aGUgZGlhbG9ndWUgYW5kIGFjdGlvbiBzZXF1ZW5jZXMKd2VyZSBnb29kLiBVbmxpa2UgbW9zdCBzY2llbmNlIGZpY3Rpb24gZmlsbXMsIHRoaXMgb25lIHdhcyBtb3N0bHkgcXVpZXQsIGFuZApub3QgYWxsIHRoYXQgbXVjaCBoYXBwZW5lZCwgZXhjZXB0IGR1cmluZyB0aGUgbGFzdCAxNSBtaW51dGVzLiBJIGRpZG4ndAp1bmRlcnN0YW5kIHdoeSBhIHVuaWNvcm4gd2FzIGluIHRoZSBtb3ZpZS4gVGhlIHZpc3VhbCBlZmZlY3RzIHdlcmUgZmFudGFzdGljLApob3dldmVyLCBhbmQgdGhlIG11c2ljYWwgc2NvcmUgYW5kIG92ZXJhbGwgbW9vZCB3YXMgcXVpdGUgaW50ZXJlc3RpbmcuCkEgZnV0dXJpc3QgTG9zIEFuZ2VsZXMgdGhhdCB3YXMgYm90aCBoaWdobHkgcG9saXNoZWQgYW5kIGFsc28gZmFsbGluZyBhcGFydApyZW1pbmRlZCBtZSBvZiAnT3V0bGFuZC4nIENlcnRhaW5seSwgdGhlIHN0eWxlIG9mIHRoZSBmaWxtIG1hZGUgdXAgZm9yCm1hbnkgb2YgaXRzIHBlZGFudGljIHBsb3QgaG9sZXMuCgpJZiB5b3Ugd2FudCB5b3VyIHNjaS1maSB0byBiZSBsYXNlcnMgYW5kIHNwYWNlc2hpcHMsICdCbGFkZXJ1bm5lcicgbWF5CmRpc2FwcG9pbnQgeW91LiBCdXQgaWYgeW91IHdhbnQgaXQgdG8gbWFrZSB5b3UgdGhpbmssIHRoaXMgbW92aWUgbWF5CmJlIHdvcnRoIHRoZSBtb25leS4KCg==
```
Comment (jerjou) [inline language/sentiment/out]: Is this file necessary?

File: language/sentiment/sentiment_analysis.py [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+
+def main(movie_review_filename):
+    '''Run a sentiment analysis request on text within a passed filename.'''
+
+    credentials = GoogleCredentials.get_application_default().create_scoped(
+        ['https://www.googleapis.com/auth/cloud-platform'])
+    http = httplib2.Http()
+    credentials.authorize(http)
+
+    service = discovery.build('language', 'v1beta1', http=http)
```
Comment (jerjou) [inline language/sentiment/sentiment_analysis.py]: You can just do:

```py
credentials = GoogleCredentials.get_application_default()
service = discovery.build('language', 'v1beta1', credentials=credentials)
```

The python client lib will take care of adding the right scope, and authorizing the http object.

File: language/sentiment/sentiment_analysis.py [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+                    'content': review_file.read(),
+                }
+            }
+        )
+        response = service_request.execute()
+
+    try:
+        polarity = response['documentSentiment']['polarity']
+        magnitude = response['documentSentiment']['magnitude']
+    except KeyError:
```
Comment (jerjou) [inline language/sentiment/sentiment_analysis.py]: Just let the `KeyError` propagate up. Error handling doesn't really provide much educational value IMO..

File: language/sentiment/sentiment_analysis.py [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+        print("The response did not contain the expected fields.")
```
Comment (jerjou) [inline language/sentiment/sentiment_analysis.py]: style: prefer single quotes over double quotes

File: language/sentiment/sentiment_analysis.py [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+    print('Sentiment: polarity of %s with magnitude of %s'
+          % (polarity, magnitude))
```
Comment (jerjou) [inline language/sentiment/sentiment_analysis.py]: Prefer using the `'string {}'.format(arg)` syntax

File: language/sentiment/sentiment_analysis.py [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+    return 0
+
+
+if __name__ == '__main__':
+    parser = argparse.ArgumentParser()
```
Comment (jerjou) [inline language/sentiment/sentiment_analysis.py]: Add a module-level docstring and set that as the `description`. See <http://go/python-sample-guide#argparse-section>

File: language/sentiment/mixed.txt (hunk 5 above)
Comment (manshreck) [inline language/sentiment/mixed.txt]: The "sentiment" dir refers to this tutorial, and I'd like to keep it named for what it is. Did you mean move the test files into a subdirectory: "sentiment/resources"? I can do that.

File: language/sentiment/out (hunk 6 above)
Comment (manshreck) [inline language/sentiment/out]: No. Will remove.

File: language/sentiment/sentiment_analysis.py (hunk 7 above)
Comment (manshreck) [inline language/sentiment/sentiment_analysis.py]: I prefer the former way even though it is more steps, because it emphasizes that you first get credentials, and then authorize them explicitly via http (and that's how I document it in the tutorial).

Comment (jonparrott) [inline language/sentiment/sentiment_analysis.py]: Please use the idiomatic way that @jerjou suggested, there's absolutely no reason to expose the http client to the user, and we're actively trying to remove httplib2. The less samples I have to clean up after that, the better.

File: language/sentiment/README.txt (hunk 4 above)
Comment (manshreck) [inline language/sentiment/README.txt]: Is this really necessary? I don't see that in a lot of other samples, and I think it will add a lot of implementation detail that's unnecessary. For complex samples, using virtualenv and a requirements.txt file is good practice, but I don't quite agree it should be always the case.

File: language/sentiment/sentiment_analysis.py (hunk 8 above)
Comment (manshreck) [inline language/sentiment/sentiment_analysis.py]: Will do, but there are competing philosophies about this. I, for one, don't like to see typical Python stack traces for known error cases, even if they aren't handled.

File: language/sentiment/README.txt (hunk 4 above)
Comment (jonparrott) [inline language/sentiment/README.txt]: Yes, every sample should use a virtualenv. Not using a virtualenv is like going out in the rain without an umbrella.

Comment (manshreck) [inline language/sentiment/README.txt]: Can you show me a README that you wish to use as the canonical README for that purpose?

Comment (jerjou) [inline language/sentiment/README.txt]: <http://go/python-samples-guide#canonical-sample> has apparently been factored away, but you can check out [python-docs-samples/storage](https://github.com/GoogleCloudPlatform/python-docs-samples/tree/master/storage) for its README

File: language/sentiment/mixed.txt (hunk 5 above)
Comment (jerjou) [inline language/sentiment/mixed.txt]: No - for all the other python samples, files used by the code for testing, etc, are in a directory named `resources`. Unless there's a good reason to deviate, having consistency between our samples is valuable. Can you update your tutorial to say "resources" instead of "sentiment"?

File: language/sentiment/sentiment_analysis.py (hunk 8 above)
Comment (jerjou) [inline language/sentiment/sentiment_analysis.py]: Yeah - I think it comes down to trading off teaching about best practices, vs keeping the signal / noise ratio high (where 'signal' is 'stuff that's relevant to the api I'm trying to learn about'). My feeling is that sample code should be as focused as is prudent on the specific thing you're trying to teach, so the user doesn't have to wade through boilerplate to get at the point. More general best practices can be covered in other venues, where they can be given the exposition they deserve.

File: language/sentiment/README.txt (hunk 2 above)
Comment (manshreck) [inline language/sentiment/README.txt]: Done

File: language/sentiment/README.txt (hunk 3 above)
Comment (manshreck) [inline language/sentiment/README.txt]: Done

File: language/sentiment/README.txt (hunk 1 above)
Comment (manshreck) [inline language/sentiment/README.txt]: Done

File: language/sentiment/README.txt (hunk 4 above)
Comment (manshreck) [inline language/sentiment/README.txt]: Done

Comment (manshreck) [inline language/sentiment/README.txt]: Also added requirements.txt

File: language/sentiment/mixed.txt (hunk 5 above)
Comment (manshreck) [inline language/sentiment/mixed.txt]: You're missing my point. All of these files are already in a "language/sentiment" directory. I'm not renaming that to "language/resources." If you want me to put these text files into a subdirectory so that they are in a "language/sentiment/resources" directory, that I can do.

File: language/sentiment/out (hunk 6 above)
Comment (manshreck) [inline language/sentiment/out]: Removed.

File: language/sentiment/sentiment_analysis.py (hunk 7 above)
Comment (manshreck) [inline language/sentiment/sentiment_analysis.py]: Done

File: language/sentiment/sentiment_analysis.py (hunk 8 above)
Comment (manshreck) [inline language/sentiment/sentiment_analysis.py]: Ackknowledged.

File: language/sentiment/sentiment_analysis.py (hunk 9 above)
Comment (manshreck) [inline language/sentiment/sentiment_analysis.py]: Done

File: language/sentiment/sentiment_analysis.py (hunk 10 above)
Comment (manshreck) [inline language/sentiment/sentiment_analysis.py]: Done

File: language/sentiment/mixed.txt (hunk 5 above)
Comment (jonparrott) [inline language/sentiment/mixed.txt]: Yes, `language/sentiment/resources` would be preferable.

File: language/sentiment/sentiment_analysis.py (hunk 11 above)
Comment (manshreck) [inline language/sentiment/sentiment_analysis.py]: Done

File: language/sentiment/mixed.txt (hunk 5 above)
Comment (manshreck) [inline language/sentiment/mixed.txt]: Done

File: language/sentiment/README.md [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+
+## Prerequisites
+
+1. Install the [Google Cloud SDK](https://cloud.google.com/sdk/), including the [gcloud tool](https://cloud.google.com/sdk/gcloud/), and [gcloud app component](https://cloud.google.com/sdk/gcloud-app).
+
+2. Setup the gcloud tool. This provides authentication to Google Cloud APIs and services.
+
+```
+$ gcloud init
+```
```
Comment (jerjou) [inline language/sentiment/README.md]: Actually, I think this won't work. Does `gcloud init` set up a service account for you? IIUC, without setting up a service account, this will attempt to use the gcloud auth, which doesn't have the api enabled.

File: language/sentiment/README.md [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+[Sentiment Analysis Tutorial](http://cloud.google.com/natural-language/docs/sentiment-tutorial)
+within the Google Cloud Natural Language API Documentation. A full walkthrough of this sample
+is located within the documentation.
+
+This sample is a simple illustration of how to construct a sentiment analysis
+request and process a response using the API.
+
+## Prerequisites
+
+1. Install the [Google Cloud SDK](https://cloud.google.com/sdk/), including the [gcloud tool](https://cloud.google.com/sdk/gcloud/), and [gcloud app component](https://cloud.google.com/sdk/gcloud-app).
```
Comment (jerjou) [inline language/sentiment/README.md]: This step isn't necessary anymore

**Comment by jerjou** [comment]
@dpebot could you kindly merge this pull request once travis has given its approval? Many thanks! Cheerio and all the best in all your endeavors!

**Comment by jonparrott** [comment]
@jerjou I'm gonna have to add that to his trigger phrases.

**Comment by jonparrott** [comment]
@dpebot merge when travis passes

#triggered.

**Comment by jerjou** [comment]
Still can't merge it :-(

**Comment by jonparrott** [comment]
@jerjou I merged your other PR, is this one still needed?

**Comment by jerjou** [comment]
Yeah - this one can be closed.
